mlrun/datastore/base.py

```
@@ -157,6 +158,176 @@ def upload(self, key, src_path):
+                if DataStore._is_directory_in_range(start_time, end_time, **kwargs):
+                    list_partition_paths_helper(
+                        paths, start_time, end_time, current_path, partion_level
+                    )
+
+        list_partition_paths_helper(
+            paths, start_time, end_time, base_path, partition_level
+        )
+        for i in range(len(paths)):
+            paths[i] = DataStore._reconstruct_path_from_base_path(base_path, paths[i])
+        return paths
+
+    @staticmethod
+    def _reconstruct_path_from_base_path(base_path: str, returned_path: str) -> str:
+        base_path = base_path.rstrip("/")
+        clean_return = returned_path.strip("/")
+
+        prefix, _, base_suffix = base_path.partition(":///")
+        schema = prefix + ":///"
+
+        return schema + clean_return
+
+    @staticmethod
+    def _clean_filters_for_partitions(
+        filters,
+        partition_keys,
+    ):
+        """
+        Remove partition keys from filters.
+
+        :param filters: (list of lists of tuples) pandas-style filters
+                Example: [[('year','=',2025),('month','=',11),('timestamp','>',ts1)]]
+        :param partition_keys: (list of str) partition columns handled via directory
+
+        :return list of list of tuples: cleaned filters
+        """
+        cleaned_filters = []
+        for group in filters:
+            new_group = [f for f in group if f[0] not in partition_keys]
+            if new_group:
+                cleaned_filters.append(new_group)
+        return cleaned_filters
+
+    @staticmethod
+    def _read_partitioned_parquet(
+        base_path,
+        start_time,
+        end_time,
+        partition_keys,
+        df_module,
+        filesystem,
+        **kwargs,
+    ):
+        """Read only the relevant partitions using pandas filters and concat."""
+        logger.debug("starting paths partition process")
+        if not partition_keys:
+            raise ValueError(
```

Need to raise an error from `mlrun.errors`. But really, this check isn't necessary. We just need to check this before this function is called (which we do).